File: LeetCode/3.java
import java.util.HashMap;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        int max = 0, p1 = 0, p2 = 0;
        HashMap<Character, Integer> window = new HashMap<Character, Integer>();
        int length = s.length();
        for (; p2 < length; p2++) {
            if (window.containsKey(s.charAt(p2)) && window.get(s.charAt(p2)) >= p1) {
                max = Math.max(max, p2 - p1);
                p1 = window.get(s.charAt(p2)) + 1;
                window.replace(s.charAt(p2), p2);
            } else {
                window.put(s.charAt(p2), p2);
            }
        }
        return Math.max(max, p2 - p1);
    }
}

File: LeetCode/3.php
<?php

class Solution
{

    /**
     * @param String $s
     * @return Integer
     */
    function lengthOfLongestSubstring($str)
    {
        $result = $p1 = $p2 = 0;
        $window = [];
        $length = strlen($str);
        for (; $p2 < $length; $p2++) {
            if (key_exists($str[$p2], $window) && $window[$str[$p2]] >= $p1) {
                $result = max($result, $p2 - $p1);
                $p1 = $window[$str[$p2]] + 1;
            }
            $window[$str[$p2]] = $p2;
        }
        return max($result, $p2 - $p1);
    }
}
